trieste/bayesian_optimizer.py

```diff
@@ -39,47 +40,69 @@
 
 
 @dataclass(frozen=True)
-class LoggingState(Generic[S]):
-    """
-    Container for the state of the optimization process in :class:`BayesianOptimizer` at each
-    optimization step.
-    """
+class Record(Generic[S]):
+    """ Container to record the state of each step of the optimization process. """
 
     datasets: Mapping[str, Dataset]
-    """ All observer data at this optimization step. """
+    """ The known data from the observer. """
 
     models: Mapping[str, TrainableProbabilisticModel]
     """ The models over the :attr:`datasets`. """
 
     acquisition_state: Optional[S]
-    """ The acquisition state after this optimization step. """
+    """ The acquisition state. """
 
 
+# this should be a generic NamedTuple, but mypy doesn't support them
+#  https://github.com/python/mypy/issues/685
 @dataclass(frozen=True)
 class OptimizationResult(Generic[S]):
-    """ Container for the final result of the optimization process. """
+    """ The final result, and the historical data of the optimization process. """
 
-    datasets: Mapping[str, Dataset]
+    final_result: Result[Record[S]]
     """
-    All data from the observer (unless :attr:`error` is populated, in which case this is the data
-    from the point at which the process was interrupted).
+    The final result of the optimization process. This contains either a :class:`Record` or an
+    exception.
     """
 
-    models: Mapping[str, TrainableProbabilisticModel]
+    history: List[Record[S]]
+    r"""
+    The history of the :class:`Record`\ s from each step of the optimization process. These
+    :class:`Record`\ s are created at the *start* of each loop, and as such will never include the
+    :attr:`final_result`.
     """
-    The models over the :attr:`datasets` (unless :attr:`error` is populated, in which case this is
-    the models from the point at which the process was interrupted). """
 
-    history: List[LoggingState[S]]
-    """ The data, models, and acquisition state at each completed optimization step. """
+    def astuple(self) -> Tuple[Result[Record[S]], List[Record[S]]]:
+        """
+        **Note:** In contrast to the standard library function :func:`dataclasses.astuple`, this
+        method does *not* deepcopy instance attributes.
+
+        :return: The :attr:`final_result` and :attr:`history` as a 2-tuple.
+        """
+        return self.final_result, self.history
+
+    def try_get_final_data(self) -> Mapping[str, Dataset]:
```

Comment:
```suggestion
    def try_get_final_datasets(self) -> Mapping[str, Dataset]:
```
for consistency throughout the code

Reply:
Or, you could change the implementation of this to assert that Dataset only has a single key:value pair and return its value (and perhaps like wise for try_get_final_model?)

Reply:
done